### PR TITLE
CSV Sniffer Function

### PR DESCRIPTION
**Pull request by pdet**
This PR introduces a stand-alone CSV sniffer function. It essentially performs the CSV sniffer on a file and returns the sniffer's dialect, column names, column types, and options.

The function takes two parameters:

1. The file path.
2. (Optional) Sample size to run the sniffer.

For example, both of these commands are valid:
``` sql
FROM sniff_csv('my_file.csv');

FROM sniff_csv('my_file.csv', sample_size = 100000);
```

The function returns the following result:
1. Delimiter (e.g., |)
2. Quote (e.g., ")
3. Escape (e.g., \)
4. NewLine Delimiter (e.g., \r\n)
5. Skip Rows (e.g., 1)
6. Has Header (e.g., True)
7. List<Struct<Column-Name:Types>> (e.g., ({'name': 'VARCHAR', 'age': 'BIGINT'})
8. Date Format (e.g., %d/%m/%Y)
9. Timestamp Format (e.g., %Y-%m-%dT%H:%M:%S.%f)
10. Prompt ready to be used in DuckDB e.g.,
```sql
FROM read_csv('data/csv/inconsistent_cells.csv', auto_detect=false, delim=',', quote='"', escape='"', new_line='\n', skip=5, header=0, columns={'column0': 'BIGINT', 'column1': 'BIGINT', 'column2': 'BIGINT', 'column3': 'BIGINT', 'column4': 'BIGINT'});
```

### CODE REVIEW

File: src/function/table/sniff_csv.cpp [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+string NewLineIdentifierToString(NewLineIdentifier identifier) {
+	switch (identifier) {
+	case NewLineIdentifier::SINGLE:
+		return "\\n";
+	case NewLineIdentifier::CARRY_ON:
+		return "\\r\\n";
+	case NewLineIdentifier::MIX:
+		return "mix";
+	case NewLineIdentifier::NOT_SET:
+		throw InternalException("NewLine Identifier must always be set after running the CSV sniffer");
```
Comment (Mytherin) [inline src/function/table/sniff_csv.cpp:94]: nit: should we throw an exception on default here instead?

File: src/function/table/sniff_csv.cpp [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+};
+
+static unique_ptr<GlobalTableFunctionState> CSVSniffInitGlobal(ClientContext &context, TableFunctionInitInput &input) {
+	return make_uniq<CSVSniffGlobalState>();
+}
+
+static unique_ptr<FunctionData> CSVSniffBind(ClientContext &context, TableFunctionBindInput &input,
+                                             vector<LogicalType> &return_types, vector<string> &names) {
+	auto result = make_uniq<CSVSniffFunctionData>();
+	result->path = input.inputs[0].ToString();
```
Comment (Mytherin) [inline src/function/table/sniff_csv.cpp:34]: Might not be that simple/important - but perhaps it makes sense to allow all parameters that we allow for `read_csv_auto` to be specified here as well, e.g. allowing the user to override the delimiter, etc?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! LGTM - some minor comments:

**Comment by pdet** [comment]
Alright, inspired by @Mytherin's comment, this PR has undergone some changes.

Most notably, I've refactored the way we handle CSV options that can be set by the system and manually by users. This allows us to have more informative error messages, to be sure never to lose variables set by the user, and to perform match checking with what we find with the sniffer.

Additionally, the `sniff_csv` function accepts almost all of the same options as the `read_csv` function. The only options not supported are those related to readers that parse multiple CSV files (this will probably change in the future), and the `auto_detect`.

**Pull request merged**